Add post about testing Quarkus apps with Citrus (#1839)
quarkusio/quarkusio.github.io: _posts/2023-11-20-testing-quarkus-with-citrus.adoc

---
layout: post
title: 'Testing Quarkus with Citrus'
date: 2023-11-20
tags: testing event-driven integration kafka messaging
synopsis: 'Explore how to verify Quarkus event-driven applications with the Citrus integration test framework.'

> author: christophd

Reviewer: you need to create an entry for this author in  `_data/authors.yaml` otherwise it shows up author-less.

Author: ah good point, thanks. let me add this

Author: Done

---
:imagesdir: /assets/images/posts/citrus

image::citrus-quarkus.png[Citrus & Quarkus,align="center"]

> This post shows how to combine Quarkus with the test framework https://citrusframework.org[Citrus] in order write automated tests for event-driven applications.

Reviewer 2: ```suggestion
This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order write automated tests for event-driven applications.
```

> This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order write automated tests for event-driven applications.

Reviewer 3: Missing 'to' after 'order': 

This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order to write automated tests for event-driven applications.

> This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order write automated tests for event-driven applications.

Reviewer 3: ```suggestion
This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order to write automated tests for event-driven applications.
```

> https://citrusframework.org[Citrus] is an Open Source Java test framework with focus on messaging and integration testing in general.

Reviewer 2: ```suggestion
https://citrusframework.org[Citrus] is an Open Source Java test framework focusing on messaging and integration testing in general.
```

The Quarkus application connects to Kafka streams as an event-driven application that produces and consumes various events (e.g. bookings, supplies, shipping events).
The processed events and their individual status are stored in a PostgreSQL database.

image::food-market-demo-application.png[Food Market App,align="center"]

The food-market application matches incoming `booking` and `supply` events and produces `shipping` and `booking-completed` events accordingly.

Each event references a product and specifies an amount as well as a price in a simple Json object structure.

[source, json]
----
{ "client": "citrus-test", "product": "Pineapple", "amount":  100, "price":  0.99 }
----

Clients create the `booking` events and at the same time suppliers will add their individual `supply` events.
The Quarkus food-market application consumes both event types and finds matching bookings and supplies.
Once a booking and a supply do match in certain criteria the application produces `booking-completed` and `shipping` events as a result.

> Last not least the booking client gets informed via email about the completed booking status.

Reviewer 2: ```suggestion
Last but not least the booking client gets informed via email about the completed booking status.
```

In a fully automated integration test we want to verify all events and their processing using real messaging communication with Kafka streams and database persistence.

== Testing the application with Citrus

The Quarkus application connects to different infrastructure (Kafka, PostgreSQL, Mail SMTP).
The automated integration test should verify the message communication, the event processing and connectivity to all components.
We will use the Citrus test framework as it provides the complete toolset for testing this kind of event-driven message-based solutions.

The first thing to do is to add Citrus to the Quarkus project.

> The most convenient way is to import the `citrus-bom`.

Reviewer: is the citrus-bom required?

how about getting the citrus-quarkus extension listed on code.quarkus.io? 

then hopefully quarkus ext add citrus-quarkus would be sufficient?

Reviewer 2: https://github.com/quarkiverse/quarkiverse/wiki/Checklist#make-your-extension-available-in-the-tooling

Reviewer 3: Or, better yet, https://hub.quarkiverse.io/checklist/#make-your-extension-available-in-the-tooling (same content as what @geoand shared, but formatted html... the indexing for that content is still a bit poor, but that's a different conversation!).

Author: +1 to list the extension on code.quarkus.io

the citrus-quarkus extension POM does not include all Citrus modules. How would the dependency resolving work in that case when there is no bom imported? 

In general the citrus-bom is not required, it only avoids to add the dependency version setting on all included `citrus-*` modules.

Reviewer 3: Hmm. @aloubyansky, can you share thoughts on this?

Reviewer 3: I wonder if the platform mechanism makes sense for this scenario: https://quarkus.io/guides/platform ... there would be a Citrus platform, which maps to the Citrus bom. Camel do this now, and there are a few smaller projects that also provide their own platform.

Reviewer 4: I am not sure about adding this extension to the platform at this point.
The text makes sense to me. AFAIU, it's either a BOM or a shared version property.

Reviewer 3: I was thinking more that we would create a new Citrus platform to hold its bom, not add it to the main Quarkus platform.

Reviewer 4: I wouldn't do it either, so far we have only one platform in our registry.

Reviewer 3: Oh, sorry, I'm using the wrong words. I think I mean platform member. 

Things like 
```
io.quarkus.platform:quarkus-kogito-bom-quarkus-platform-descriptor
io.quarkus.platform:quarkus-camel-bom-quarkus-platform-descriptor
io.quarkus.platform:quarkus-amazon-services-bom-quarkus-platform-descriptor
```

The test needs to wait for the product event to be processed completely before sending the booking event.

In Citrus we are able to add this waiting state very easily.

.Wait for object to be created in persistence layer
[source, java]
----
Product product = new Product("Watermelon");
t.when(send()
    .endpoint(products)
    .message().body(marshal(product)));

t.then(repeatOnError()
    .condition((i, context) -> i > 25)
    .autoSleep(500)
    .actions(
        sql().dataSource(dataSource)
            .query()
            .statement("select count(id) as found from product where product.name='%s'"
                    .formatted(product.getName()))
            .validate("found", "1"))
);
----

After the product event has been sent we use the `repeatOnError()` test action.
In combination with an `autoSleep` and a max retry count setting the action periodically polls the database for the created product.
This makes sure that we do not continue with the test until the new product has been properly stored to the database.

The database interaction in Citrus comes with the `citrus-sql` module and enables you to verify any SQL result set.

> TIP: Quarkus is able to inject the `dataSource` that is being used to connect to the PostreSQL database. This also works when Quarkus uses the PostreSQL Testcontainers infrastructure in the test. Just use the `@Inject` annotation in your test and reference the datasource in the Citrus `sql()` test action.

Reviewer 3: ```suggestion
TIP: Quarkus is able to inject the `dataSource` that is being used to connect to the PostgreSQL database. This also works when Quarkus uses the PostreSQL Testcontainers infrastructure in the test. Just use the `@Inject` annotation in your test and reference the datasource in the Citrus `sql()` test action.
```
